version.js

```diff
@@ -1,92 +1,291 @@
-import { warning } from "@actions/core";
+import { warning, debug, isDebug } from "@actions/core";
 import { context } from "@actions/github";
-import { parse as parseSemver } from "semver";
+import { SemVer } from "semver";
 import { Octokit } from "octokit";
 
+// Only write debug messages when the RUNNER_DEBUG environment variable is set.
+// This reduces noise in tests.
+const localDebug = isDebug() ? debug : () => {};
+
 /**
  * Calculate the version to use for the current build.
  * @param {any} fetch
  * @param {typeof context} context
  */
 export async function calculateVersion(context) {
-  // Are we building a tagged release?
-  if (context.eventName === "push" && context.ref.startsWith("refs/tags/v")) {
-    // Get the version from the tag
-    const version = context.ref.replace("refs/tags/", "");
-    const parsed = parseSemver(version); // Ensure it's a valid semver version
-    if (parsed === null) {
-      throw new Error(`Invalid tag version: ${version}`);
+  const eventName = context.eventName;
+  const ref = context.ref;
+  const sha = context.sha;
+  const defaultBranch = context.payload?.repository?.default_branch;
+
+  localDebug(`event_name: ${eventName}`);
+  localDebug(`ref: ${ref}`);
+  localDebug(`sha: ${sha}`);
+  localDebug(`repository.default_branch: ${defaultBranch}`);
+
+  if (eventName === "push" && ref.startsWith("refs/tags/")) {
+    localDebug(`Tag pushed: ${ref}`);
+    return calculateTagVersion(ref);
+  }
+
+  if (eventName === "push" && ref.startsWith("refs/heads/")) {
+    // push events only
+    const headCommitTimestamp = context.payload?.head_commit?.timestamp;
+    const headCommitMessage = context.payload?.head_commit?.message;
+    localDebug(`head_commit.timestamp: ${headCommitTimestamp}`);
+    localDebug(`head_commit.message: ${headCommitMessage}`);
+
+    const branchName = ref.replace("refs/heads/", "");
+    const asVersion = tryParseVersionBranch(branchName);
+    if (asVersion !== undefined) {
+      localDebug(`Version branch pushed: ${branchName}`);
+      const baseVersion = new SemVer(`${asVersion}.0.0`);
+      return alphaVersion(baseVersion, headCommitTimestamp);
+    }
+    if (branchName === defaultBranch) {
+      localDebug(`Default branch pushed: ${defaultBranch}`);
+      const previousRelease = await getLatestReleaseVersion(context.repo);
+      const increment = await getIncrementType(headCommitMessage);
+      const nextVersion = previousRelease.inc(increment);
+      return alphaVersion(nextVersion, headCommitTimestamp);
     }
-    return parsed.version;
+    localDebug(`Branch pushed: ${branchName}`);
+    const previousRelease = await getLatestReleaseVersion(context.repo);
+    const nextVersion = previousRelease.inc("minor");
```

Comment:
Theoretically, it could be a patch version, right? Is "minor" simply an assumption?

Reply:
Yup, minor is just an assumption - given most of our releases are minor versions. This is the same as pulumictl assumed.

In theory we could allow customisation by using labels e.g. `needs-release/minor` or `needs-release/patch`, though I think a more robust form would be to require each PR to indicate if it's a bug fix or new feature which could then be used to automatically bump the version on the next release based on the maximum of the PRs merged since the previous release ... and could tie into this action too to produce an accurate pre-release ... though making it any more accurate is probably mostly vanity or at best avoiding minor confusion for anyone trawling through pre-releases.